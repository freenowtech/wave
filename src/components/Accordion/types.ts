export interface AccordionProps {
    /**
     * Heading of the Accordion section
     */
    heading?: string;
    /**
     * Description of the section
     */
    description?: string;
    /**
     * Extra text below description
     */
    info?: string;
    /**
     * Text on the button
     */
    buttonLabel?: string;
    /**
     * UI style
     * @default default
     */
    variant?: 'compact' | 'default';
    /**
     * Render accordion initially opened
     * @default false
     */
    defaultExpanded?: boolean;
    /**
     * Callback, runs after the accordion is opened
     */
    onExpand?: () => void;
    /**
     * Callback, runs after the accordion is closed
     */
    onCollapse?: () => void;
}
