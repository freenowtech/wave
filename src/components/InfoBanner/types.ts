import { ReactNode } from 'react';
import { BoxProps } from '../Box/Box';

export type InfoBannerVariants = 'info' | 'success' | 'warning' | 'error';

export interface CardProps extends BoxProps {
    children: ReactNode;
    variant?: InfoBannerVariants;
    emphasized?: boolean;
}

export interface BoxWithVariant extends BoxProps {
    variant: InfoBannerVariants;
    emphasized: boolean;
}
