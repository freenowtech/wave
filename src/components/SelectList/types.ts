import { Props } from 'react-select';
import { MarginProps, WidthProps } from 'styled-system';
import { ClassNameProps } from '../../utils/extractProps';

export type Variant = 'boxed' | 'bottom-lined';
export type Size = 'medium' | 'small';

interface SelectListProps extends Props, ClassNameProps, WidthProps, MarginProps {
    variant?: Variant;
    size?: Size;
    inverted?: boolean;
    error?: boolean;
    label?: string;
    inputId?: string;
}

export { SelectListProps };
