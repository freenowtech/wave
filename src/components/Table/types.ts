type SortingDirection = 'DESC' | 'ASC';

export { SortingDirection };
