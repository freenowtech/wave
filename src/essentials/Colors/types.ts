import { Join, Leaves } from '../../utils/types';

export type HSL = `hsl(${number}, ${number}%, ${number}%)` | `hsla(${number}, ${number}%, ${number}%, ${number})`;

export type Color = HSL | 'transparent';

export type SemanticColorsSchema = {
    transparent: 'transparent';
    white: Color;
    black: Color;
    background: {
        page: {
            default: Color;
            'elevation-1': Color;
            'elevation-2': Color;
            'elevation-3': Color;
        };
        backdrop: Color;
        // for big areas
        surface: {
            primary: {
                default: Color;
                emphasized: Color;
            };
            neutral: {
                default: Color;
                faded: Color;
                emphasized: Color;
            };
            info: {
                default: Color;
                active: Color;
                faded: Color;
                emphasized: Color;
            };
            success: {
                default: Color;
                emphasized: Color;
            };
            warning: {
                default: Color;
                emphasized: Color;
            };
            danger: {
                faded: Color;
                default: Color;
                emphasized: Color;
            };
        };
        // for form elements and other small areas
        element: {
            primary: {
                default: Color;
                emphasized: Color;
            };
            neutral: {
                default: Color;
                emphasized: Color;
            };
            disabled: {
                faded: Color;
                default: Color;
            };
            accent: {
                default: Color;
                faded: Color;
                emphasized: Color;
            };
            info: {
                default: Color;
                emphasized: Color;
            };
            success: {
                default: Color;
            };
            warning: {
                default: Color;
            };
            danger: {
                default: Color;
                faded: Color;
                emphasized: Color;
            };
        };
    };
    border: {
        neutral: {
            faded: Color;
            default: Color;
            emphasized: Color;
        };
        accent: {
            faded: Color;
            default: Color;
        };
        info: {
            banner: Color;
            faded: Color;
            default: Color;
        };
        success: {
            banner: Color;
            faded: Color;
            default: Color;
        };
        warning: {
            banner: Color;
            default: Color;
        };
        danger: {
            banner: Color;
            faded: Color;
            default: Color;
        };
        focus: Color;
        disabled: Color;
    };
    foreground: {
        disabled: Color;
        focus: Color;
        primary: Color;
        'on-background': {
            primary: Color;
            accent: Color;
            info: Color;
            success: Color;
            warning: Color;
            danger: Color;
            disabled: Color;
            neutral: Color;
        };
        neutral: {
            faded: Color;
            default: Color;
            emphasized: Color;
        };
        accent: {
            default: Color;
            emphasized: Color;
        };
        info: {
            faded: Color;
            default: Color;
        };
        success: {
            default: Color;
        };
        warning: {
            default: Color;
        };
        danger: {
            default: Color;
            emphasized: Color;
        };
    };
    logo: {
        free: Color;
        now: Color;
        subtitle: Color;
    };
    shadow: {
        default: Color;
    };
};

type SemanticColorToken = Join<Leaves<SemanticColorsSchema>, '-'>;
type SemanticColorTokenCssVariable = `--wave-s-color-${SemanticColorToken}`;

type SemanticHslComponentsToken = `${SemanticColorToken}-hsl`;
type HslComponentsTokenCssVariable = `--wave-s-color-${SemanticHslComponentsToken}`;

export type SemanticToken = SemanticColorToken | SemanticHslComponentsToken;
export type SemanticColorCssVariable = SemanticColorTokenCssVariable | HslComponentsTokenCssVariable;

export type ReadCssColorVariable =
    | `var(${SemanticColorTokenCssVariable})`
    | `var(${SemanticColorTokenCssVariable}, ${string})`;
export type ReadCssVariable = `var(${SemanticColorCssVariable})` | `var(${SemanticColorCssVariable}, ${string})`;

export type ComponentSemanticTokens = {
    [key: string]: ReadCssVariable | ComponentSemanticTokens;
};
