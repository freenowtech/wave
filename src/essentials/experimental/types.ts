import { Color } from '../Colors/types';

export type Palette = {
    [key in 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 97]: Color;
};

export type ExtendedPalette = Palette & {
    [key in 0 | 99 | 100]: Color;
};

export type ColorPaletteSchema = {
    neutral: ExtendedPalette;
    marooned: Palette;
    red: Palette;
    green: Palette;
    blue: Palette;
    orange: Palette;
    yellow: Palette;
};

export type SemanticColorsSchema = {
    // Accent
    accent: Color;
    'on-accent': Color;
    // Interactive
    interactive: Color;
    'on-interactive': Color;
    'interactive-container': Color;
    'on-interactive-container': Color;
    // Neutrals
    surface: Color;
    'on-surface': Color;
    'surface-variant': Color;
    'on-surface-variant': Color;
    outline: Color;
    'outline-variant': Color;
    // Inverse
    'inverse-surface': Color;
    'inverse-on-surface': Color;
    'inverse-on-surface-variant': Color;
    // Informative
    negative: Color;
    'negative-container': Color;
    'on-negative-container': Color;
    positive: Color;
    'positive-container': Color;
    'on-positive-container': Color;
};
