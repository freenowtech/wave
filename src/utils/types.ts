/**
 * Returns a union of all possible paths to leaves in an object.
 *
 * @example
 * type Foo = {
 *   bar: {
 *     baz: string;
 *   },
 *   qux: number;
 * }
 *
 * type FooLeaves = Leaves<Foo>;
 * // ['bar', 'baz'] | ['qux']
 */
export type Leaves<T> = {
    [K in keyof T]: T[K] extends Record<string | number, unknown> ? [K, ...Leaves<T[K]>] : [K];
}[keyof T];

/**
 * Joins a union of strings with a separator.
 *
 * @example
 * type Foo = ['bar', 'baz'] | ['a', 'b', 'c'];
 *
 * type Joined = Join<Foo>;
 * // 'bar-baz' | 'a-b-c'
 */
export type Join<T extends unknown[], U extends string = '-'> = T extends [infer F, ...infer R]
    ? R['length'] extends 0
        ? `${F & (string | number)}`
        : `${F & (string | number)}${U}${Join<R, U>}`
    : never;
