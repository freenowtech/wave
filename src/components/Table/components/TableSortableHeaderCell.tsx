import React, { FC } from 'react';
import styled from 'styled-components';
import { Box } from '../../Box/Box';
import { SortingIndicator } from './SortingIndicator';
import { TableHeaderCell, TableHeaderCellProps } from './TableHeaderCell';
import type { SortingDirection } from '../types';

const TableHeaderCellWithPointer = styled(TableHeaderCell)`
    cursor: pointer;
    user-select: none;
`;

type TableSortableHeaderCellProps = TableHeaderCellProps & {
    active: boolean;
    field: string;
    direction: SortingDirection;
    onSortChange: (field: string, direction: string) => void;
};

const TableSortableHeaderCell: FC<TableSortableHeaderCellProps> = ({
    active,
    direction,
    children,
    field,
    onSortChange,
    ...rest
}: TableSortableHeaderCellProps) => (
    <TableHeaderCellWithPointer
        onClick={() => onSortChange(field, direction)}
        aria-sort={active ? (direction === 'ASC' ? 'ascending' : 'descending') : 'none'}
        {...rest}
    >
        <Box display="inline-flex" alignItems="center">
            {children}
            <Box mr="0.25rem" />
            <Box flexShrink="0" display="inline-flex" alignItems="center">
                <SortingIndicator direction={active ? direction : 'NONE'} />
            </Box>
        </Box>
    </TableHeaderCellWithPointer>
);

export { TableSortableHeaderCell, TableSortableHeaderCellProps };
