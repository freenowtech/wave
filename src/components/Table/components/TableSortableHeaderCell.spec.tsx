import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { SortingDirection } from '../types';
import { TableSortableHeaderCell } from './TableSortableHeaderCell';

const handleSortChange = jest.fn();

const renderComponent = (active = false, direction = 'ASC') =>
    render(
        <table>
            <thead>
                <tr>
                    <TableSortableHeaderCell
                        active={active}
                        field="address"
                        direction={direction as SortingDirection}
                        onSortChange={handleSortChange}
                    >
                        Address
                    </TableSortableHeaderCell>
                </tr>
            </thead>
        </table>
    );

describe('TableSortableHeaderCell', () => {
    it('should indicate the ASC direction', () => {
        renderComponent(true, 'ASC');

        expect(screen.getByText('Address', { exact: false })).toMatchSnapshot();
    });

    it('should indicate the DESC direction', () => {
        renderComponent(true, 'DESC');

        expect(screen.getByText('Address', { exact: false })).toMatchSnapshot();
    });

    it('should have aria-sort none when inactive', () => {
        renderComponent();

        expect(screen.getByText('Address', { exact: false }).parentElement).toHaveAttribute('aria-sort', 'none');
    });

    it('should have aria-sort ascending when active and direction is ASC', () => {
        renderComponent(true, 'ASC');

        expect(screen.getByText('Address', { exact: false }).parentElement).toHaveAttribute('aria-sort', 'ascending');
    });

    it('should have aria-sort descending when active and direction is DESC', () => {
        renderComponent(true, 'DESC');

        expect(screen.getByText('Address', { exact: false }).parentElement).toHaveAttribute('aria-sort', 'descending');
    });

    it('should call the callback with the right arguments', () => {
        renderComponent(true, 'ASC');

        userEvent.click(screen.getByText('Address', { exact: false }));

        expect(handleSortChange).toHaveBeenCalledWith('address', 'ASC');
    });
});
