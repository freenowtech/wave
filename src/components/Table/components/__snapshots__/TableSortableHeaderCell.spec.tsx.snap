// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableSortableHeaderCell should indicate the ASC direction 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  margin-right: 0.25rem;
}

.c2 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  class="c0"
  display="inline-flex"
>
  Address
  <div
    class="c1"
  />
  <div
    class="c2"
    display="inline-flex"
  >
    <svg
      color="#001E3E"
      fill="none"
      height="20"
      viewBox="0 0 20 20"
      width="20"
    >
      <path
        d="M10 15.8334L7.83494 13.3334L5.66987 10.8334L10 10.8334L14.3301 10.8334L12.1651 13.3334L10 15.8334Z"
        fill="#C6CDD4"
      />
      <path
        d="M10 4.16663L7.83494 6.66663L5.66987 9.16663L10 9.16663L14.3301 9.16663L12.1651 6.66663L10 4.16663Z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;

exports[`TableSortableHeaderCell should indicate the DESC direction 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  margin-right: 0.25rem;
}

.c2 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  class="c0"
  display="inline-flex"
>
  Address
  <div
    class="c1"
  />
  <div
    class="c2"
    display="inline-flex"
  >
    <svg
      color="#001E3E"
      fill="none"
      height="20"
      viewBox="0 0 20 20"
      width="20"
    >
      <path
        d="M10 15.8334L7.83494 13.3334L5.66987 10.8334L10 10.8334L14.3301 10.8334L12.1651 13.3334L10 15.8334Z"
        fill="currentColor"
      />
      <path
        d="M10 4.16663L7.83494 6.66663L5.66987 9.16663L10 9.16663L14.3301 9.16663L12.1651 6.66663L10 4.16663Z"
        fill="#C6CDD4"
      />
    </svg>
  </div>
</div>
`;
